### ISSUE TITLE
GRPOTrainer silently uses near-greedy decoding when temperature=1.0 (transformers >= 4.50 + Qwen2.5)

### ISSUE THREAD

**Issue by timwei0801** [huggingface/trl#5783]
## Summary

`GRPOTrainer` silently runs near-greedy decoding when `GRPOConfig.temperature=1.0` (the default) is used with any Qwen2.5-family model since `transformers >= 4.50.0`. This collapses GRPO rollouts to near-duplicates, making the advantage signal degenerate (`std(R) ≈ 0`) and silently breaking the algorithm's exploration mechanism. Training runs to completion with no error.

## Reproduction

Same SFT-trained LoRA adapter, same prompt, `temperature=1.0`:

* **Direct `model.generate()`** (32 samples) → 7 distinct outputs in 13 parses ✓ diverse
* **TRL GRPOTrainer** (8 rollouts, `num_generations=8`) → 4 of 8 completions have **identical md5 hashes** ❌

Minimal repro (Qwen2.5-VL-3B):
```python
import torch
from transformers import GenerationConfig

# === BUGGY (matches TRL behavior) ===
gen_config = GenerationConfig(
    temperature=1.0, top_p=1.0, top_k=None,
    repetition_penalty=1.0, do_sample=True, max_new_tokens=512,
)
out_buggy = model.generate(input_ids=batch_8_duplicated_rows, generation_config=gen_config)
# → 4 unique md5s / 8 rows

# === FIXED ===
out_fixed = model.generate(
    input_ids=batch_8_duplicated_rows, generation_config=gen_config,
    use_model_defaults=False,  # ← the fix
)
# → 8 unique md5s / 8 rows
```

Buggy path prints this warning (easily missed):
```
`generation_config` default values have been modified to match model-specific defaults:
{'max_length': 128000, 'temperature': 1e-06, 'repetition_penalty': 1.05, 'bos_token_id': 151643}.
```

## Root cause

`transformers.generation.utils.GenerationMixin._prepare_generation_config` (introduced in transformers 4.50.0) silently overrides user-set fields in `GenerationConfig` that equal the **global defaults** with the model's bundled `generation_config.json` values, unless the caller passes `use_model_defaults=False` to `model.generate(...)`.

Qwen2.5-VL-3B-Instruct's bundled `generation_config.json` ships with:
```json
{
  "temperature": 1e-06,
  "top_p": 0.001,
  "top_k": 1,
  "repetition_penalty": 1.05,
  "transformers_version": "4.50.dev0"
}
```

`GRPOTrainer.__init__` (around `grpo_trainer.py:564-579`) builds `self.generation_config` directly from `GRPOConfig` defaults (1.0, 1.0, None, 1.0) — every field happens to be a global default — and `_generate_single_turn` (around `grpo_trainer.py:1292`) calls `unwrapped_model.generate(generation_config=self.generation_config, ...)` **without** passing `use_model_defaults=False`. All four sampling fields silently revert to Qwen's near-greedy values inside `model.generate()`.

Effect on GRPO: `num_generations` rollouts per prompt become near-duplicates → `std(R) ≈ 0` in the advantage formula → effectively no gradient signal for exploration.

## Why this matters

We discovered this while investigating "Red Agent mode collapse" in an adversarial training setup. After ~6 weeks of trying every 2025-recommended exploration remedy — DAPO clip-higher (arXiv:2503.14476), GAPO group-frequency reward (arXiv:2511.12596), Active Attacks defender refresh (arXiv:2509.21947), NGRPO advantage calibration (arXiv:2509.18851) — none of them worked because the rollouts were never actually diverse to begin with. After applying the fix below, within-group diversity is restored on the same training setup.

Any HuggingFace user running GRPO on a Qwen-family model with default `GRPOConfig.temperature=1.0` since `transformers 4.50.0` (released ~2025-01) is silently affected.

## Proposed fix

Pass `use_model_defaults=False` in `_generate_single_turn` (and `_generate_single_turn_vllm` doesn't need it — vLLM bypasses this code path):

```python
prompt_completion_ids = unwrapped_model.generate(
    **generate_inputs,
    generation_config=self.generation_config,
    disable_compile=True,
    use_model_defaults=False,   # ← prevent silent override
)
```

This documents intent: TRL has carefully constructed `self.generation_config` from user-provided `GRPOConfig` and the user expects those values to be used as-is, not silently merged with the model's bundled config.

## Workaround until upstream fix

Subclass `GRPOTrainer` (or any of its descendants) and bump default-equal sampling fields by 1 ulp in `__init__`:

```python
def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs)
    gc = self.generation_config
    if float(getattr(gc, "temperature", 1.0)) == 1.0:
        gc.temperature = 1.0000001          # bit-equivalent sampling
    if getattr(gc, "top_p", None) == 1.0:
        gc.top_p = 0.9999999
    if getattr(gc, "top_k", None) is None:
        gc.top_k = 0                         # 0 = disabled (same as None)
    if float(getattr(gc, "repetition_penalty", 1.0)) == 1.0:
        gc.repetition_penalty = 1.0000001
```

Because none of these values equal the global defaults anymore, `_prepare_generation_config` leaves them alone. Sampling distribution unchanged at the floating-point level.

## Affected versions

- TRL: 0.18 through at least 0.24.0 (any version calling `model.generate(generation_config=...)` in the generation path)
- transformers: ≥ 4.50.0
- Confirmed on: `Qwen/Qwen2.5-VL-3B-Instruct`, `Qwen/Qwen2.5-VL-7B-Instruct-unsloth-bnb-4bit`
- Likely affects: Qwen2.5-Instruct text-only and other recent Qwen variants whose bundled `generation_config.json` declares `transformers_version >= 4.50.0` and ships with non-default sampling values

## Environment

- `trl==0.24.0`
- `transformers==4.57.6`
- `vllm==0.15.1` (TRL's vLLM rollout path is **not** affected — it bypasses HF GenerationConfig and uses `SamplingParams` directly)
- Qwen2.5-VL-3B-Instruct (Unsloth bf16 + LoRA, also unsloth-bnb-4bit)

**Comment by timwei0801** [comment]
**Update / self-correction**: after submitting this I checked the codebase more carefully and found this is a duplicate of [transformers#42762](https://github.com/huggingface/transformers/issues/42762) (filed by @albertvillanova), fixed in transformers v5.0.0 by [transformers#42702](https://github.com/huggingface/transformers/pull/42702). TRL `main` has also already added a workaround for `transformers < 5` via [`_override_model_generation_config`](https://github.com/huggingface/trl/blob/main/trl/models/utils.py#L154) inside `unwrap_model_for_generation(generation_kwargs=...)`. Apologies for the duplicate report.

That said, the reason I (and presumably other users) are still hitting this in May 2026 is the version pinning interaction:

- TRL 0.24.0 (the latest tagged release at the time of writing) **does not** contain the `_override_model_generation_config` workaround — it landed on `main` after the 0.24.0 tag
- transformers >=5.0 unbreaks the underlying merge, but transformers 5.x has its own forward-compat issues with current vLLM 0.15.x (e.g. `Qwen2VLImageProcessorFast.max_pixels=None` crash on cold start), so production users are pinned back to transformers 4.5x
- → users on the latest stable TRL + latest stable vLLM combo end up on transformers 4.5x without the TRL main workaround, silently running near-greedy GRPO on Qwen2.5 with no error

Would the TRL team consider one of:

1. **A patch release to 0.24.x** that backports `_override_model_generation_config` (it's a self-contained ~40 LOC addition to `trl/models/utils.py`, no public API change). 0.24.x is widely deployed and many users won't track `main`.
2. **A release-notes warning / FAQ entry** explicitly flagging "if you're on TRL 0.24.x + transformers 4.5x, your GRPO sampling may silently revert to model defaults" with a pointer to the workaround we've shown above.

Happy to send a PR for (1) if the maintainers think it's worth doing — the change has no public API impact and is already battle-tested on `main`.

Sorry again for the noise. Closing-status I'll leave to the maintainers; the issue may still be useful as a discoverable record for other 0.24.x users.

**Comment by smqd19** [comment]
Sure, I've seen this issue before. It's a common problem when using near-greedy decoding with models like Qwen2.5. The default `temperature` of 1.0 tends to collapse the rollouts, leading to near-duplicates. This can happen due to the model's internal defaults being applied, which can override the `generation_config` settings.

In our production environment, we've found that explicitly setting `use_model_defaults=False` can help. This forces the model to use the provided `generation_config` instead of falling back to its internal defaults. This small tweak ensures that the exploration mechanism in GRPO is preserved.

Here's a quick example of how you can apply this fix:

```python
out_fixed = model.generate(
    input_ids=batch_8_duplicated_rows, generation_config=gen_config,
    use_model_defaults=False,  # ← the fix
)
```

This should give you the 8 unique md5s you're looking for.

**Comment by albertvillanova** [comment]
Hey @timwei0801 — thanks for the unusually thorough writeup, and for tracking the root cause down yourself. 

You're right that this is handled on `main`: `_override_model_generation_config` (PR #4647) temporarily forces the model's `generation_config` to our training values during rollout, sidestepping the transformers `_prepare_generation_config` merge. It first shipped in v0.27.0 and is in every release since (current latest is v1.7.1). The key detail for your setup: that workaround is version-guarded to activate precisely when `transformers < 5.0.0` (see https://github.com/huggingface/trl/blob/cfef01a9188240dc6f830f3d7abb10075891efed/trl/models/utils.py#L173-L180), so you don't need to move to transformers 5 to get it — just bumping TRL to >= 0.27.0 restores rollout diversity on your pinned transformers 4.5x + vLLM 0.15.x combo. GRPO wires it in here: https://github.com/huggingface/trl/blob/cfef01a9188240dc6f830f3d7abb10075891efed/trl/trainer/grpo_trainer.py#L1721-L1725. On the 0.24.x backport: we roll forward rather than maintaining patch branches on old minors, so the recommendation is to upgrade rather than backport. 

Closing as fixed — please re-open if you still see the collapse on >= 0.27.0.
